operator/src/main/java/oracle/kubernetes/operator/Main.java

```diff
@@ -467,36 +467,64 @@ Step createCRDPresenceCheck() {
 
     @Override
     public NextAction apply(Packet packet) {
-      V1CustomResourceDefinition existingCrd = delegate.getCrdReference().get();
-      if (existingCrd != null) {
-        warnedOfCrdAbsence = false;
+      return doNext(new CallBuilder().readCustomResourceDefinitionAsync(
+              KubernetesConstants.CRD_NAME, createReadResponseStep(getNext())), packet);
+    }
+  }
+
+  // If CRD read succeeds, wait until the CRD has webhook.
+  // Otherwise if CRD read fails due to permissions error, list the domains to indirectly check if the CRD is installed.
+  ResponseStep<V1CustomResourceDefinition> createReadResponseStep(Step next) {
+    return new ReadResponseStep(next);
+  }
+
+  class ReadResponseStep extends DefaultResponseStep<V1CustomResourceDefinition> {
+    ReadResponseStep(Step next) {
+      super(next);
+    }
+
+    @Override
+    public NextAction onSuccess(
+            Packet packet, CallResponse<V1CustomResourceDefinition> callResponse) {
+      V1CustomResourceDefinition existingCrd = callResponse.getResult();
+
+      if (!existingCrdContainsConversionWebhook(existingCrd)) {
+        LOGGER.info(MessageKeys.WAIT_FOR_CRD_INSTALLATION, CRD_DETECTION_DELAY);
+        return doDelay(createCRDPresenceCheck(), packet, CRD_DETECTION_DELAY, TimeUnit.SECONDS);
+      } else {
         return doNext(packet);
       }
-      return doNext(new CallBuilder().listDomainAsync(getOperatorNamespace(),
-          new CrdPresenceResponseStep(getNext())), packet);
+    }
+
+    private boolean existingCrdContainsConversionWebhook(V1CustomResourceDefinition existingCrd) {
+      return Optional.ofNullable(existingCrd).map(crd -> crd.getSpec()).map(spec -> spec.getConversion())
+              .map(c -> c.getStrategy().equalsIgnoreCase(WEBHOOK)).orElse(false);
+    }
+
+    @Override
+    protected NextAction onFailureNoRetry(Packet packet, CallResponse<V1CustomResourceDefinition> callResponse) {
+      return isNotAuthorizedOrForbidden(callResponse)
+              ? doNext(new CallBuilder().listDomainAsync(getOperatorNamespace(),
+                new CrdPresenceResponseStep(getNext())), packet) : super.onFailureNoRetry(packet, callResponse);
     }
   }
 
-  // on failure, aborts the processing.
-  class CrdPresenceResponseStep extends DefaultResponseStep<DomainList> {
+  // on failure, waits for the CRD to be installed.
+  private class CrdPresenceResponseStep extends DefaultResponseStep<DomainList> {
 
     CrdPresenceResponseStep(Step next) {
       super(next);
     }
 
     @Override
     public NextAction onSuccess(Packet packet, CallResponse<DomainList> callResponse) {
```

Comment:
With the removal of warnedOfCrdAbsence then you don't need to define onSuccess() at all as it just calls the super method.

Reply:
Removed.